### PR TITLE
DashedVMobject reworked

### LINKED ISSUE #1803: Bug in DashedVMobject: the last dash in a circle is too short
## Description of bug / unexpected behavior
Going to see the example displayed at https://docs.manim.community/en/stable/reference/manim.mobject.types.vectorized_mobject.DashedVMobject.html#manim.mobject.types.vectorized_mobject.DashedVMobject

We see this:
image

Look carefully at the second row:
image

You can see that the last dash is too short, and the gap between dashes is to big between the last and the first (if dashes are constructed in the trigonometric direction).

## Expected behavior
All dashes in a circle have the same lenght. All gaps between dashes are the same.

## How to reproduce the issue
Code (same as in the exemple):

```py
from manim import *

class DashedVMobjectExample(Scene):
    def construct(self):
        r = 0.5

        top_row = VGroup()  # Increasing num_dashes
        for dashes in range(2, 12):
            circ = DashedVMobject(Circle(radius=r, color=WHITE), num_dashes=dashes)
            top_row.add(circ)

        middle_row = VGroup()  # Increasing positive_space_ratio
        for ratio in np.arange(1 / 11, 1, 1 / 11):
            circ = DashedVMobject(
                Circle(radius=r, color=WHITE), positive_space_ratio=ratio
            )
            middle_row.add(circ)

        sq = DashedVMobject(Square(1.5, color=RED))
        penta = DashedVMobject(RegularPolygon(5, color=BLUE))
        bottom_row = VGroup(sq, penta)

        top_row.arrange(buff=0.4)
        middle_row.arrange()
        bottom_row.arrange(buff=1)
        everything = VGroup(top_row, middle_row, bottom_row).arrange(DOWN, buff=1)
        self.add(everything)
```

### PR DESCRIPTION
**Pull request by sparshg**
## Changelog / Overview

`positive_space_ratio` renamed to `dashed_ratio`
`dash_spacing` parameter removed from `DashedLine`
Other bugs fixed

## Explanation for Changes

Closes #1803 
Also other improvements
### Before
Also note the right end of the white dashed line during this animation (increasing positive space ratio from 0 to 1)

https://user-images.githubusercontent.com/43041139/126474585-3dd5230a-8945-4fb9-b2a4-9b68c209ce3c.mp4

Note the inconsistent gaps in the middle line

Before

### After

https://user-images.githubusercontent.com/43041139/126474577-989d275c-1d53-46f1-a814-58bf11c40fcc.mp4

After

<details>
<summary>Code</summary>

```py
class Test(Scene):
    def construct(self):
        v = ValueTracker(0)
        d = DashedVMobject(Line()).add_updater(
            lambda m: m.become(
                DashedVMobject(Line(), 3, v.get_value()).shift(2 * RIGHT)
            )
        )
        l = DashedVMobject(Circle()).add_updater(
            lambda m: m.become(
                DashedVMobject(Circle(), 3, v.get_value()).shift(2 * LEFT)
            )
        )
        self.add(Line().shift(UP * 0.1 + 2 * RIGHT), d, l)
        self.play(v.animate.set_value(1), run_time=4)
```

```py
class DashedVMobjectExample(Scene):
    def construct(self):
        r = 0.5

        top_row = VGroup()  # Increasing num_dashes
        for dashes in range(2, 12):
            circ = DashedVMobject(Circle(radius=r, color=WHITE), num_dashes=dashes)
            top_row.add(circ)

        middle_row = VGroup()  # Increasing positive_space_ratio
        for ratio in np.arange(1 / 11, 1, 1 / 11):
            circ = DashedVMobject(
                Circle(radius=r, color=WHITE), positive_space_ratio=ratio
            )
            middle_row.add(circ)

        sq = DashedVMobject(Square(1.5, color=RED))
        penta = DashedVMobject(RegularPolygon(5, color=BLUE))
        bottom_row = VGroup(sq, penta)

        top_row.arrange(buff=0.4)
        middle_row.arrange()
        bottom_row.arrange(buff=1)
        everything = VGroup(top_row, middle_row, bottom_row).arrange(DOWN, buff=1)
        self.add(everything)
```
</details>

[Documentation Reference](https://manimce--####.org.readthedocs.build)

## Testing Status

Everything now perfect and symmetrical

```
=================================== FAILURES ===================================
_____________ [doctest] manim.mobject.geometry.DashedLine.get_end ______________

DashedLine().get_end()

Expected:
    array([0.99871795, 0.        , 0.        ])
Got:
    array([1., 0., 0.])
```
= Perfect
```
_________ [doctest] manim.mobject.geometry.DashedLine.get_last_handle __________

DashedLine().get_last_handle()
Expected:
    array([0.98205128, 0.        , 0.        ])
Got:
    array([0.98333333, 0.        , 0.        ])
```
= Symmetrical to DashedLine().get_start_handle() which returns -0.9833

## Checklist
- [ ] I have read the [Contributing Guidelines](https://docs.manim.community/en/latest/contributing.html)
- [ ] I have written a descriptive PR title (see top of PR template for examples)
- [ ] I have written a changelog entry for the PR or deem it unnecessary
- [ ] My new functions/classes either have a docstring or are private
- [ ] My new functions/classes have [tests](https://github.com/ManimCommunity/manim/wiki/Testing) added and (optional) examples in the docs
- [ ] My new documentation builds, looks correctly formatted, and adds no additional build warnings

## Reviewer Checklist
- [ ] The PR title is descriptive enough
- [ ] The PR is labeled correctly
- [ ] The changelog entry is completed if necessary
- [ ] Newly added functions/classes either have a docstring or are private
- [ ] Newly added functions/classes have [tests](https://github.com/ManimCommunity/manim/wiki/Testing) added and (optional) examples in the docs
- [ ] Newly added documentation builds, looks correctly formatted, and adds no additional build warnings

### CODE REVIEW

File: manim/mobject/types/vectorized_mobject.py [hunk 1]
```diff
@@ -2107,30 +2107,22 @@ def __init__(
...
-            # End points of the unit interval for division
-            alphas = np.linspace(0, 1, num_dashes + 1)
-
-            # This determines the length of each "dash"
-            full_d_alpha = 1.0 / num_dashes
-            partial_d_alpha = full_d_alpha * ps_ratio
-
-            # Shifts the alphas and removes the last dash
-            # to give closed shapes even spacing
+            # Assuming total length is 1
```
Comment (huguesdevimeux) [inline manim/mobject/types/vectorized_mobject.py:2113]: I don't understand this, what is the total length ?

Comment (sparshg) [reply manim/mobject/types/vectorized_mobject.py:2113]: total length is like the vmobject's perimeter which we assume as 1 map it from 0% to 100% of the vmobject's subcurve... same as alpha interpolating from 0 to 1 as it was before the changes... but it was really complicated to understand and fix for me

Comment (huguesdevimeux) [reply manim/mobject/types/vectorized_mobject.py:2113]: I get the idea, but I don't understand how it's linked to this line. It's just the space ratio divided by the number of dashes, no ?

File: manim/mobject/types/vectorized_mobject.py [hunk 2]
```diff
@@ -2107,30 +2107,22 @@ def __init__(
...
+            dash_len = p / n
```
Comment (huguesdevimeux) [inline manim/mobject/types/vectorized_mobject.py:2114]: I think this is wrong; this should be to compute the void length and not the dash len. 

As a visualization, look at the second example you attached with the PR : 
image

So, I think you just mixed but dash_len and void_len.

**Review by huguesdevimeux — CHANGES_REQUESTED** [review]
- I think you made a small mistake, check my comment 
- Please add a graphical test for that. 
Thanks :D

File: manim/mobject/types/vectorized_mobject.py (hunk 2 above)
Comment (sparshg) [reply manim/mobject/types/vectorized_mobject.py:2114]: I still dont see any problem, as p increases, the dash_len (p/n) is increasing and void_len that is (1-p)/n is decreasing

File: manim/mobject/types/vectorized_mobject.py (hunk 1 above)
Comment (sparshg) [reply manim/mobject/types/vectorized_mobject.py:2113]: Yes, we are basically dividing our unit length into p : (1-p), and then splitting it into n parts, hence p/n and (1-p)/n are dash_len and void_len respectively

**Comment by huguesdevimeux** [comment]
I still strongly think the name `positive_space_ratio` should be change to `dashed_ratio` (just moving the conversation we had on discord here)

**Comment by sparshg** [comment]
okay so I was looking at DashedLine now because I wanted to see what dash_spacing does, but it doesn't seem to affect anything, and there's calculate_ps_ratio() function and.. what? lol I think this is broken too, let me try fixing

**Comment by sparshg** [comment]
ok so dash_spacing was redundant, after doing my changes I saw opengl version also removed dash_spacing lol

**Review by huguesdevimeux — APPROVED** [review]

File: manim/mobject/geometry.py [hunk 3]
```diff
@@ -1091,8 +1091,8 @@ class DashedLine(Line):
         The length of each individual dash of the line.
     dash_spacing : Optional[:class:`float`]
```
Comment (hydrobeam) [inline manim/mobject/geometry.py:1092]: Since `dash_spacing` doesn't isn't used at all now, it should probably be removed imo

**Review by hydrobeam — APPROVED** [review]
The changes seem solid to me🎉! 

There should probably be a mention in the changelog that `dashed_ratio` has replaced `positive_space_ratio` (since these are technically breaking changes)  + maybe a more descriptive PR title 🤔.

**Comment by sparshg** [comment]
Removed the leftovers, everything seems good now

**Pull request merged**